controllers/machine_controller.go

```diff
@@ -83,6 +89,34 @@ func (r *MachineReconciler) SetupWithManager(mgr ctrl.Manager, options controlle
 		return errors.Wrap(err, "failed setting up with a controller manager")
 	}
 
+	// 'UpdateFunc' and 'CreateFunc' used to judge if an event about the object is
+	// what we want. If that is true, the event will be processed by the reconciler.
+	p := predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			oldCluster := e.ObjectOld.(*clusterv1.Cluster)
+			newCluster := e.ObjectNew.(*clusterv1.Cluster)
+			return oldCluster.Spec.Paused && !newCluster.Spec.Paused
+		},
+		CreateFunc: func(e event.CreateEvent) bool {
+			if _, ok := e.Meta.GetAnnotations()[clusterv1.PausedAnnotation]; !ok {
+				return false
+			}
+			return true
+		},
+	}
+
+	// Watch Deployments and trigger Reconciles for objects
+	// mapped from the Deployment in the event
+	err = controller.Watch(
+		&source.Kind{Type: &clusterv1.Cluster{}},
+		&handler.EnqueueRequestsFromMapFunc{
+			ToRequests: handler.ToRequestsFunc(r.reconcileRequests),
+		},
+		p)
```

Comment:
maybe just define the predicate here? it could provide a nice pattern for the future, we can also inline the comments

Reply:
It might be good to define the predicate funcs similar to the way we are defining the mapping methods, so that we could also add unit testing around them.

Reply:
that works too, they're small enough for now that they can be inlined, also don't wanna block this in any ways

Reply:
inlined for now, happy to revisit if needed